src/tests/ftest/cart/util/cart_logtest.py

```diff
@@ -576,73 +581,85 @@ def _check_pid_from_log_file(self,
             raise WarningMode()
 #pylint: enable=too-many-branches,too-many-nested-blocks
 
-    def rpc_reporting(self, pid):
-        """RPC reporting for RPC state machine, for mutiprocesses"""
-        op_state_counters = {}
-        c_states = {}
-        c_state_names = set()
+class rpc_reporting():
+    """Class for reporting a summary of RPC states"""
 
-        # Use to convert from descriptor to opcode.
-        current_opcodes = {}
+    known_functions = frozenset({'crt_hg_req_send',
+                                 'crt_hg_req_destroy',
+                                 'crt_rpc_complete',
+                                 'crt_rpc_priv_alloc',
+                                 'crt_rpc_handler_common',
+                                 'crt_req_send',
+                                 'crt_hg_req_send_cb'})
 
-        for line in self._li.new_iter(pid=pid):
-            rpc_state = None
-            opcode = None
+    def __init__(self):
 
-            function = getattr(line, 'function', None)
-            if not function:
-                continue
-            if line.is_new_rpc():
-                rpc_state = 'ALLOCATED'
-                opcode = line.get_field(-4)
-                if opcode == 'per':
-                    opcode = line.get_field(-8)
-            elif line.is_dereg_rpc():
-                rpc_state = 'DEALLOCATED'
-            elif line.endswith('submitted.'):
-                rpc_state = 'SUBMITTED'
-            elif function == 'crt_hg_req_send' and \
-                 line.get_field(-6) == ('sent'):
-                rpc_state = 'SENT'
-
-            elif line.is_callback():
-                rpc = line.descriptor
-                rpc_state = 'COMPLETED'
-                result = line.get_field(13).split('(')[0]
-                c_state_names.add(result)
-                opcode = current_opcodes[line.descriptor]
-                try:
-                    c_states[opcode][result] += 1
-                except KeyError:
+        self._op_state_counters = {}
+        self._c_states = {}
+        self._c_state_names = set()
+        self._current_opcodes = {}
 
-                    c_states[opcode] = Counter()
-                    c_states[opcode][result] += 1
-            else:
-                continue
+    def add_line(self, line):
+        """Parse a output line"""
+
+        try:
+            if line.function not in self.known_functions:
+                return
+        except AttributeError:
+            return
```

Comment:
line.function may not exist.  I try and write code like "if hasattr(line, 'function'):..." but python really does like the "try it and catch the failure" style of programming which I'm learning to live this, this is an example of that.  This line is also performance critical, so loading it once is probably a good thing.

Reply:
This one example is interesting. It shows that python `try..except` is slightly more efficient than a simple `if` statement in cases where the `if` would have evaluated to `True`. Otherwise, the `try..except` performance hit is probably negligible.
I can live with this :)

Reply:
Also, line.filename is processes via a __getattr__ callback, so checking for it, then reading it would be two calls to https://github.com/daos-stack/daos/blob/master/src/tests/ftest/cart/util/cart_logparse.py#L178-L200